helix-core/src/transaction.rs

```diff
@@ -513,6 +576,43 @@ impl Transaction {
         Self::change(doc, selection.iter().map(f))
     }
 
+    pub fn change_by_selection_ignore_overlapping<T>(
+        doc: &Rope,
+        selection: &Selection,
+        mut create_change: impl FnMut(&Range, &ChangeSet) -> Option<(Change, T)>,
+        mut process_applied_change: impl FnMut(&Range, T),
+    ) -> (Transaction, usize) {
+        let mut last_selection_idx = None;
+        let mut new_primary_idx = 0;
+        let mut idx = 0;
+        let mut applied_changes = 0;
+        let process_change = |change: Option<_>| {
+            if let Some((range, data)) = change {
+                last_selection_idx = Some(applied_changes);
+                process_applied_change(range, data);
+                applied_changes += 1;
+            }
+            if idx == selection.primary_index() {
+                new_primary_idx = last_selection_idx.unwrap_or(0);
+            }
+            idx += 1;
+        };
+        let mut selection = selection.iter();
+        let transaction = Self::change_ignore_overlapping(
+            doc,
+            selection.len(),
+            |changset| {
+                selection.next().map(|range| {
+                    let (change, data) = create_change(range, changset)?;
+                    Some((change, (range, data)))
+                })
+            },
+            process_change,
+        );
+        // map the transaction trough changes
+        (transaction, new_primary_idx)
```

Comment:
Why is it lossy?

Reply:
a selection is sorted and overlapping ranges are merged on creation. In this case it's fine tough because the selection is already a valid selection so removing some ranges won't change that so I did change back to a selection afterall